src/main/java/nextstep/security/filter/VirtualFilterChain.java

```diff
@@ -0,0 +1,37 @@
+package nextstep.security.filter;
+
+import jakarta.servlet.Filter;
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.ServletRequest;
+import jakarta.servlet.ServletResponse;
+
+import java.io.IOException;
+import java.util.List;
+
+public class VirtualFilterChain implements FilterChain {
+
+    private final FilterChain originalChain;
+
+    private final List<Filter> additionalFilters;
+
+    private final int size;
+
+    private int currentPosition = 0;
+
+    public VirtualFilterChain(FilterChain originalChain, List<Filter> additionalFilters) {
+        this.originalChain = originalChain;
+        this.additionalFilters = additionalFilters;
+        this.size = additionalFilters.size();
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response) throws IOException, ServletException {
+        if (currentPosition == size) {
+            originalChain.doFilter(request, response);
+        } else {
+            Filter nextFilter = additionalFilters.get(currentPosition++);
+            nextFilter.doFilter(request, response, this);
+        }
```

Comment:
```suggestion
        if (currentPosition == size) {
            originalChain.doFilter(request, response);
            return;
        }
        Filter nextFilter = additionalFilters.get(currentPosition++);
        nextFilter.doFilter(request, response, this);    
```

코드 가독성을 생각해본다면 size가 동일할때 return된다고 명시하는 것이 루프가 "종료"됨을 좀 더 명확하게 표현해줄 수 있을 것 같네요 :)